src/NDItk/base/Keyword.hpp

```diff
@@ -0,0 +1,75 @@
+#ifndef NJOY_NDITK_BASE_KEYWORD
+#define NJOY_NDITK_BASE_KEYWORD
+
+// system includes
+#include <optional>
+#include <regex>
+#include <string>
+#include <iostream>
+
+// other includes
+#include "tools/Log.hpp"
+
+namespace njoy {
+namespace NDItk {
+namespace base {
+
+/**
+ *  @brief A class for NDI keywords
+ *
+ *  This class allows us to decompose
+ */
+class Keyword {
+
+  /* fields */
+  std::string keyword_;
+  std::optional< std::string > subtype_ = std::nullopt;
```

Comment:
Looking further down, I don't follow. Is keyword the keyword without the _, and subtype is the keyword with the _? And then particle would be the particle type. In other words:

keyword = num_grps
subtype = num_grps_
particle = 1001

Reply:
Figured it out. Is the whole point of Keyword to allow for particle keywords? Is the overhead not a big deal?

Reply:
keyword = num_grps_1001
subtype = num_grps
particle = 1001

Yes, I introduced keyword to deal with these cases. I'll do the same with the prompt, total and delayed part of the nubar keys using an enumerator instead of an int. It should feature regex to extract those things at the end.